R/attributes.R

```diff
@@ -539,28 +539,36 @@ vertex.attributes <- function(graph, index = V(graph)) {
     any(names(value) == "") || any(duplicated(names(value)))) {
     stop("Value must be a named list with unique names")
   }
-  if (any(sapply(value, length) != length(index))) {
+  if (!all(sapply(value, length) == length(index))) {
     stop("Invalid attribute value length, must match number of vertices")
   }
 
   if (!missing(index)) {
     index <- as_igraph_vs(graph, index)
 
-    if (any(duplicated(index)) || any(is.na(index))) {
+    if ((anyDuplicated(index) > 0) || (anyNA(index) > 0)) {
       stop("Invalid vertices in index")
     }
   }
 
+  index_number_not_vcount <- (length(index) != vcount(graph))
+  index_not_vertices_indices <- (!all(index %in% V(graph)))
   if (!missing(index) &&
-    (length(index) != vcount(graph) || any(index != V(graph)))) {
-    vs <- V(graph)
-    for (i in seq_along(value)) {
-      tmp <- value[[i]]
+    (index_number_not_vcount || index_not_vertices_indices)) {
+
+    make_value_correct_vector <- function(value_item, vcount) {
+      tmp <- value_item
       length(tmp) <- 0
```

Comment:
This does something different based on whether `value_item` is a vector or a list.

``` r
thing <- list(1, 2, 4)
length(thing) <- 0
length(thing) <- 4
thing
#> [[1]]
#> NULL
#> 
#> [[2]]
#> NULL
#> 
#> [[3]]
#> NULL
#> 
#> [[4]]
#> NULL

stuff <- c(1, 2, 4)
length(stuff) <- 0
length(stuff) <- 4
stuff
#> [1] NA NA NA NA
```

<sup>Created on 2024-04-02 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>